# Alex Germick - Final
github link:

In [59]:
import pandas as pd
import numpy as np

from keras.datasets import imdb
from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Flatten
from keras.layers import Dense, Dropout, Activation
from keras.layers.embeddings import Embedding

from keras.layers import GlobalMaxPooling1D
from keras.layers.convolutional import Conv1D
from keras.layers.convolutional import MaxPooling1D

- Load data

In [60]:
(x_train, y_train), (x_test, y_test) = imdb.load_data(path="imdb.npz",
                                                      num_words=5000,
                                                      skip_top=0,
                                                      maxlen=None,
                                                      seed=123,
                                                      start_char=1,
                                                      oov_char=2,
                                                      index_from=3)

In [61]:
print(x_test.shape)
print(x_train.shape)

(25000,)
(25000,)


- Build NN Model

In [62]:
model = Sequential()

max_features = 5000
maxlen = 400
batch_size = 32
embedding_dims = 50
filters = 250
kernel_size = 3
hidden_dims = 250
epochs = 2

x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)

model.add(Embedding(max_features,
                    embedding_dims,
                    input_length=maxlen))
model.add(Dropout(0.2))

# we add a Convolution1D, which will learn filters
# word group filters of size filter_length:
model.add(Conv1D(filters,
                 kernel_size,
                 padding='valid',
                 activation='relu',
                 strides=1))
# we use max pooling:
model.add(GlobalMaxPooling1D())

# We add a vanilla hidden layer:
model.add(Dense(hidden_dims))
model.add(Dropout(0.2))
model.add(Activation('relu'))

# We project onto a single unit output layer, and squash it with a sigmoid:
model.add(Dense(1))
model.add(Activation('sigmoid'))

model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

In [63]:
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/2
25000/25000 [==============================] - 295s 12ms/step - loss: 0.4104 - acc: 0.7958 - val_loss: 0.3030 - val_acc: 0.8715
Epoch 2/2
25000/25000 [==============================] - 299s 12ms/step - loss: 0.2333 - acc: 0.9060 - val_loss: 0.2545 - val_acc: 0.8941


Train on 25000 samples, validate on 25000 samples
Epoch 1/2
25000/25000 [==============================] - 295s 12ms/step - loss: 0.4104 - acc: 0.7958 - val_loss: 0.3030 - val_acc: 0.8715
Epoch 2/2
25000/25000 [==============================] - 299s 12ms/step - loss: 0.2333 - acc: 0.9060 - val_loss: 0.2545 - val_acc: 0.8941
<keras.callbacks.History at 0x130ec4e0e80>